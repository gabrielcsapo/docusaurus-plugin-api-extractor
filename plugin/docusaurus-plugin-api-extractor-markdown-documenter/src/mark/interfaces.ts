import { ApiItem, ApiModel, Excerpt } from '@microsoft/api-extractor-model';
import {
  DocCodeSpan,
  DocFencedCode,
  DocLinkTag,
  DocNode,
  DocParagraph,
  DocPlainText,
  DocSection,
  TSDocConfiguration
} from '@microsoft/tsdoc';
import { IndentedWriter } from './builders/indented-builder';
import { DocEmphasisSpan } from './nodes/doc-emphasis-span';
import { DocFrontmatter } from './nodes/doc-frontmatter';
import { DocHeading } from './nodes/doc-heading';
import { DocNoteBox } from './nodes/doc-notebox';
import { DocTable } from './nodes/doc-table';
import { DocTableCell } from './nodes/doc-table-cell';
import { DocTableRow } from './nodes/doc-table-row';

export interface IEmphasisOptions {
  bold?: boolean;
  italic?: boolean;
}

export type NextPage = (apiItem: ApiItem) => void;

export type SectionBuilder = () => void;

export interface IEmitterOptions {
  contextApiItem: ApiItem | undefined;
  onGetFilenameForApiItem: (apiItem: ApiItem) => string | undefined;
}

export type YamlList =
  | string
  | number
  | boolean
  // eslint-disable-next-line @rushstack/no-new-null
  | null
  // eslint-disable-next-line @rushstack/no-new-null
  | { [key: string]: null }
  | { [key: string]: YamlList };

export interface IDocumenterDelegate {
  outputFolder: string;
  apiModel: ApiModel;
  configureTSDoc?(configuration: TSDocConfiguration): TSDocConfiguration;
  writeNode?(writeCtx: IWriteNodeContext): void;
  prepareFrontmatter?(fileName: string, pageTitle: string): YamlList;
  writePage?(b: IMarkdownDelegateContext): void;
}

export interface IEmitterContext {
  writer: IndentedWriter;
  insideTable: boolean;
  boldRequested: boolean;
  italicRequested: boolean;
  writingBold: boolean;
  writingItalic: boolean;
  options: IEmitterOptions;
}

export interface IWriteNodeContext {
  docNode: DocNode;
  context: IEmitterContext;
  docNodeSiblings: boolean;
  writeNode(docNode: DocNode, context: IEmitterContext, docNodeSiblings: boolean): void;
}

export interface IInternalDocumenterDelegate {
  outputFolder: string;
  apiModel: ApiModel;
  configureTSDoc(configuration: TSDocConfiguration): TSDocConfiguration;
  prepareFrontmatter(fileName: string, pageTitle: string): YamlList;
  writePage(b: IMarkdownDelegateContext): void;
  writeNode(ctx: IWriteNodeContext): void;
}

export interface IMarkdownDelegateContext {
  primitives: IPrimitiveBuilders;
  sections: ISectionBuilders;
  currentPage: DocSection;
  apiItem: ApiItem;
  tsDocConfiguration: TSDocConfiguration;
  append(docNode: DocNode): void;
  next(apiItem: ApiItem): void;
}

/**
 * Responsible for building primitive pieces of a markdown document
 * @public
 */
export interface IPrimitiveBuilders {
  /**
   * Creates a node represented a fenced off snippet of code
   * @param code a code snippet that will be in the fenced output
   * @param language the name of the langauge for the code snippet
   */
  code(code: string, language: string): DocFencedCode;
  /**
   * Creates a node for an inline snippet of code
   * @param code code snippet that will be inlined
   */
  codeSpan(code: string): DocCodeSpan;
  /**
   * Creates a table cell node specifically for the description of a member of a class or interface
   * @param apiItem {@link @microsoft/api-extractor-model#ApiItem | ApiItem}
   */
  descriptionCell(apiItem: ApiItem): DocTableCell;
  /**
   * Creates a node that represents either bold or italic text
   * @param options
   * @param children
   */
  emphasis(options: IEmphasisOptions, children: DocNode[]): DocEmphasisSpan;

  /**
   * Creates a paragraph node that contains linked references to other objects and interfaces
   * @param excerpt
   */
  excerpt(excerpt: Excerpt): DocParagraph;

  /**
   * Creates a heading based on the "@framworkItemType" annotation
   * @param apiItem {@link @microsoft/api-extractor-model#ApiItem | ApiItem}
   */
  frameworkItemTypeHeading(apiItem: ApiItem): DocHeading;
  /**
   * Creates a node that represents frontmatter written in YAML
   * @param list
   */
  frontmatter(list: YamlList): DocFrontmatter;
  /**
   * Creates a heading node for the passed string
   * @param heading
   */
  heading(heading: string): DocHeading;
  /**
   * Creates a link to a specific destination. Links will be made relative to the outputFolder
   * @param linkText
   * @param destination
   */
  link(linkText: string, destination: string): DocLinkTag;
  /**
   * Generates a table cell for a typescript modifier e.g. public / static / protected etc
   * @param apiItem {@link @microsoft/api-extractor-model#ApiItem | ApiItem}
   */
  modiferCell(apiItem: ApiItem): DocTableCell;
  /**
   * Creates a table cell based on the contents of the "@modulePath" annotation
   * @param apiItem {@link @microsoft/api-extractor-model#ApiItem | ApiItem}
   */
  modulePathCell(apiItem: ApiItem): DocTableCell;
  /**
   * Returns a level 2 heading based on the contents of the "@modulePath" annotation
   * @param apiItem {@link @microsoft/api-extractor-model#ApiItem | ApiItem}
   */
  modulePathHeading(apiItem: ApiItem): DocHeading | undefined;
  /**
   * Creates a node which will be a callout box in markdown
   * @param children
   */
  noteBox(children: DocNode[]): DocNoteBox;
  /**
   * Creates a node that represents a paragraph
   * @param children
   */
  paragraph(children: DocNode[]): DocParagraph;
  /**
   * Generates a table cell for a specific property of a class or interface
   * @param apiItem {@link @microsoft/api-extractor-model#ApiItem | ApiItem}
   */
  propertyTypeCell(apiItem: ApiItem): DocTableCell;
  /**
   * Creates a new section node
   */
  section(): DocSection;
  /**
   * Creates a node representing a table cell
   * @param children
   */
  tableCell(children: DocNode[]): DocTableCell;
  /**
   * Creates a node representing a table row
   * @param children
   */
  tableRow(children: DocTableCell[]): DocTableRow;
  /**
   * Creates a table with headings
   * @param headings
   */
  table(headings: string[]): DocTable;
  /**
   * Creates a node represting a chunk of plain text.
   * @param text
   */
  text(text: string): DocPlainText;
}

export interface ISectionBuilders {
  /**
   * Appends a table to the current section that will contain all the information about the class that is currently being documented
   */
  classTable: SectionBuilder;
  /**
   * Appends a table that will contain all the information about the interface that is currently being documented
   */
  interfaceTable: SectionBuilder;
  /**
   * Appends a section that contains all the heritage types for the given object currently being documented
   */
  heritageTypes: SectionBuilder;
  /**
   * Appends a table that contains all the information about the packages
   */
  modelTable: SectionBuilder;
  /**
   * Appends a section containing information about any decorators attached to a class
   */
  decorators: SectionBuilder;
  /**
   * Appends a section containing information about deprecated items for the object currently being documented
   */
  deprecated: SectionBuilder;
  /**
   * Appends a section containing the contents of the "@remarks" block associated with the object currently being documented
   */
  remarks: SectionBuilder;
  /**
   * Appends a section containing all members of a package or namespace
   */
  packageOrNamespace: SectionBuilder;
  /**
   * Appends the breadcrumb trail
   */
  breadcrumb: SectionBuilder;
  /**
   * Appends a heading for the object currently being documented
   */
  pageHeading: SectionBuilder;
  /**
   * Appends any warnings regarding APIs that have been annotated with "@\beta"
   */
  betaWarning: SectionBuilder;
  /**
   * Appends a table containing all the members of an enum
   */
  enumTable: SectionBuilder;
  /**
   * Appends frontmatter to the section
   */
  frontmatter: SectionBuilder;
  /**
   * Appends a table containing all the information about the parameters for the object currently being documented
   */
  parameterTable: SectionBuilder;
  /**
   * Appends information about errors that will be thrown by the object currently being documented
   */
  throws: SectionBuilder;
}
