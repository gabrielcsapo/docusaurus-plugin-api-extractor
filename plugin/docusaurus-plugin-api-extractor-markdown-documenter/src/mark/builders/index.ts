import { ApiModel } from '@microsoft/api-extractor-model';
import { DocSection, TSDocConfiguration } from '@microsoft/tsdoc';
import { NextPage } from './interfaces';
import { PrimitiveBuilders } from './primitive-builders';
import { SectionBuilders } from './section-builders';

export type Body = (builders: SectionBuilders) => void;

export type BodyBuilder = (body: Body) => DocSection;

export function createPageBuilder(
  configuration: TSDocConfiguration,
  outputFolder: string,
  apiModel: ApiModel,
  next: NextPage
): BodyBuilder {
  return (body: Body): DocSection => {
    const section: DocSection = new DocSection({ configuration });
    const primitiveBuilders: PrimitiveBuilders = new PrimitiveBuilders(configuration, apiModel);
    const sectionBuilders: SectionBuilders = new SectionBuilders(
      primitiveBuilders,
      section,
      apiModel,
      outputFolder,
      next
    );

    body(sectionBuilders);
    return section;
  };
}
