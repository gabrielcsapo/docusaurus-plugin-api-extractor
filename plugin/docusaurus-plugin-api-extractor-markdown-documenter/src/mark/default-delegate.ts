import { ApiItem, ApiItemKind, ApiModel } from '@microsoft/api-extractor-model';
import { TSDocConfiguration } from '@microsoft/tsdoc';
import { parse } from 'path';
import {
  IDocumenterDelegate,
  IInternalDocumenterDelegate,
  IMarkdownDelegateContext,
  IWriteNodeContext,
  YamlList
} from './interfaces';

export class InternalDelegate implements IInternalDocumenterDelegate {
  public outputFolder: string;
  public apiModel: ApiModel;
  private _delegate: IDocumenterDelegate;

  public constructor(delegate: IDocumenterDelegate) {
    this._delegate = delegate;
    this.apiModel = delegate.apiModel;
    this.outputFolder = delegate.outputFolder;
  }

  public configureTSDoc(currentConfiguration: TSDocConfiguration): TSDocConfiguration {
    if (this._delegate.configureTSDoc) {
      return this._delegate.configureTSDoc(currentConfiguration);
    }

    return currentConfiguration;
  }

  public prepareFrontmatter(fileName: string, pageTitle: string): YamlList {
    if (this._delegate.prepareFrontmatter) {
      return this._delegate.prepareFrontmatter(fileName, pageTitle);
    }
    const { name: id } = parse(fileName);
    const slug: string | undefined = id === 'index' ? '/' : undefined;

    const list: YamlList = {
      id,
      hide_title: true,
      title: pageTitle,
      // eslint-disable-next-line @typescript-eslint/naming-convention
      custom_edit_url: null
    };

    if (slug) {
      list.slug = slug;
    }

    return list;
  }

  public writeNode(ctx: IWriteNodeContext): void {
    if (this._delegate.writeNode) {
      this._delegate.writeNode(ctx);
    } else {
      throw new Error(`Unsupported DocNode kind ${ctx.docNode.kind}`);
    }
  }

  public writePage(ctx: IMarkdownDelegateContext): void {
    if (this._delegate.writePage) {
      this._delegate.writePage(ctx);
    } else {
      const { sections: b, apiItem } = ctx;
      b.frontmatter();
      b.breadcrumb();
      b.pageHeading();
      b.betaWarning();
      b.deprecated();
      b.decorators();
      b.heritageTypes();

      let appendRemarks: boolean = true;

      switch (apiItem.kind) {
        case ApiItemKind.Class:
        case ApiItemKind.Interface:
        case ApiItemKind.Namespace:
        case ApiItemKind.Package:
          b.remarks();
          appendRemarks = false;
          break;
      }

      switch (apiItem.kind) {
        case ApiItemKind.Class:
          b.classTable();
          break;
        case ApiItemKind.Enum:
          b.enumTable();
          break;
        case ApiItemKind.Interface:
          b.interfaceTable();
          break;
        case ApiItemKind.Constructor:
        case ApiItemKind.ConstructSignature:
        case ApiItemKind.Method:
        case ApiItemKind.MethodSignature:
        case ApiItemKind.Function:
          b.parameterTable();
          b.throws();
          break;
        case ApiItemKind.Namespace:
          b.packageOrNamespace();
          break;
        case ApiItemKind.Model:
          b.modelTable();
          break;
        case ApiItemKind.Package:
          b.packageOrNamespace();
          break;
        case ApiItemKind.Property:
        case ApiItemKind.PropertySignature:
          break;
        case ApiItemKind.TypeAlias:
          break;
        case ApiItemKind.Variable:
          break;
        default:
          throw new Error(`Unsupported API item kind: ${apiItem.kind}`);
      }

      if (appendRemarks) {
        b.remarks();
      }
    }
  }
}

export class DefaultDelegate implements IDocumenterDelegate {
  public apiModel: ApiModel;
  public outputFolder: string;

  public constructor(apiModel: ApiModel, oututPath: string) {
    this.apiModel = apiModel;
    this.outputFolder = oututPath;
  }

  public prepareFrontmatter(fileName: string, pageTitle: string): YamlList {
    const { name: id } = parse(fileName);
    const slug: string | undefined = id === 'index' ? '/' : undefined;

    const list: YamlList = {
      id,
      hide_title: true,
      title: pageTitle,
      // eslint-disable-next-line @typescript-eslint/naming-convention
      custom_edit_url: null
    };

    if (slug) {
      list.slug = slug;
    }

    return list;
  }
}
