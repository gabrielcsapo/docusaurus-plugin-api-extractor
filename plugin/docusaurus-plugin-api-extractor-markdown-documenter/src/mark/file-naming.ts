import { ApiItem, ApiParameterListMixin } from '@microsoft/api-extractor-model';

const BAD_FILE_NAME_CHARS: RegExp = /[^a-z0-9_\-\.]/gi;
const INVALID_NAME_CHARS: RegExp = /[^A-Za-z0-9\-_\.]/;

export function getSafeFilenameForName(name: string): string {
  return name.replace(BAD_FILE_NAME_CHARS, '_').toLowerCase();
}

export function getConciseSignature(apiItem: ApiItem): string {
  if (ApiParameterListMixin.isBaseClassOf(apiItem)) {
    return apiItem.displayName + '(' + apiItem.parameters.map((x) => x.name).join(', ') + ')';
  }
  return apiItem.displayName;
}

interface IParsedNamed {
  scope: string;
  unscopedName: string;
}

export function getParsedName(packageName: string): IParsedNamed {
  const parsedName: IParsedNamed = {
    scope: '',
    unscopedName: ''
  };

  let input: string = packageName;

  if (input === null || input === undefined) {
    throw new Error('The package name must not be null or undefined');
  }
  // Rule from npmjs.com:
  // "The name must be less than or equal to 214 characters. This includes the scope for scoped packages."
  if (packageName.length > 214) {
    throw new Error('The package name cannot be longer than 214 characters');
  }

  if (input[0] === '@') {
    const indexOfScopeSlash: number = input.indexOf('/');
    if (indexOfScopeSlash <= 0) {
      throw new Error(`Error parsing "${packageName}": The scope must be followed by a slash`);
    }
    parsedName.scope = input.substring(0, indexOfScopeSlash);
    input = input.substring(indexOfScopeSlash + 1);
  }

  parsedName.unscopedName = input;

  if (parsedName.scope === '@') {
    throw new Error(`Error parsing "${packageName}": The scope name cannot be empty`);
  }

  if (parsedName.unscopedName === '') {
    throw new Error(`The package name cannot be empty`);
  }

  if (parsedName.unscopedName[0] === '.' || parsedName.unscopedName[0] === '_') {
    throw new Error(`The package name "${packageName}" starts with an invalid character`);
  }

  const nameWithoutScopeSymbols: string =
    (parsedName.scope ? parsedName.scope.slice(1, -1) : '') + parsedName.unscopedName;

  const match: RegExpMatchArray | null = nameWithoutScopeSymbols.match(INVALID_NAME_CHARS);

  if (match) {
    throw new Error(`The package name "${packageName}" contains an invalid character: "${match[0]}"`);
  }

  return parsedName;
}
