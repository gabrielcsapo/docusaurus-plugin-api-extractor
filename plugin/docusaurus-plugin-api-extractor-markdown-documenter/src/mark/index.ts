export { StandardMarkdownDocumenter } from './standard-markdown-documenter';
export { IDocumenterDelegate, IMarkdownDelegateContext, YamlList, IWriteNodeContext } from './interfaces';
