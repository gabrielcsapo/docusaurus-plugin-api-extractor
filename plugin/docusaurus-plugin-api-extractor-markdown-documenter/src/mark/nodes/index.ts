import { DocNodeKind, TSDocConfiguration } from '@microsoft/tsdoc';
import { DocEmphasisSpan } from './doc-emphasis-span';
import { DocHeading } from './doc-heading';
import { DocNoteBox } from './doc-notebox';
import { DocTable } from './doc-table';
import { DocTableCell } from './doc-table-cell';
import { DocTableRow } from './doc-table-row';
import { CustomDocNodeKind } from './doc-types';

export class CustomDocNodes {
  private static _configuration: TSDocConfiguration | undefined;
  public static get configuration(): TSDocConfiguration {
    if (CustomDocNodes._configuration === undefined) {
      const configuration: TSDocConfiguration = new TSDocConfiguration();

      configuration.docNodeManager.registerDocNodes('@micrososft/api-documenter', [
        { docNodeKind: CustomDocNodeKind.EmphasisSpan, constructor: DocEmphasisSpan },
        { docNodeKind: CustomDocNodeKind.Heading, constructor: DocHeading },
        { docNodeKind: CustomDocNodeKind.NoteBox, constructor: DocNoteBox },
        { docNodeKind: CustomDocNodeKind.Table, constructor: DocTable },
        { docNodeKind: CustomDocNodeKind.TableCell, constructor: DocTableCell },
        { docNodeKind: CustomDocNodeKind.TableRow, constructor: DocTableRow }
      ]);

      configuration.docNodeManager.registerAllowableChildren(CustomDocNodeKind.EmphasisSpan, [
        DocNodeKind.PlainText,
        DocNodeKind.SoftBreak
      ]);

      configuration.docNodeManager.registerAllowableChildren(DocNodeKind.Section, [
        CustomDocNodeKind.Heading,
        CustomDocNodeKind.NoteBox,
        CustomDocNodeKind.Table
      ]);

      configuration.docNodeManager.registerAllowableChildren(DocNodeKind.Paragraph, [
        CustomDocNodeKind.EmphasisSpan
      ]);

      CustomDocNodes._configuration = configuration;
    }
    return CustomDocNodes._configuration;
  }
}
