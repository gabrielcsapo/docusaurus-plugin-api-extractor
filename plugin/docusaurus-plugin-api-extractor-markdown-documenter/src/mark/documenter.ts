import {
  ApiClass,
  ApiEnum,
  ApiInterface,
  ApiItem,
  ApiItemKind,
  ApiModel,
  ApiNamespace,
  ApiPackage,
  ApiParameterListMixin
} from '@microsoft/api-extractor-model';
import { StringBuilder, DocSection } from '@microsoft/tsdoc';
import path from 'path';

import { CustomDocNodes } from './nodes';
import { StandardMarkdownEmitter } from './standard-markdown-emitter';
import { BodyBuilder, createPageBuilder } from './builders';
import { getFilenameForApiItem, getLinkFilenameForApiItem } from './builders/utils';
import { ISections } from './builders/interfaces';

export class Documenter {
  private _emitter: StandardMarkdownEmitter;
  private _apiModel: ApiModel;
  private _outputFolder: string;
  private _pages: Record<string, string> = {};
  private _pageBuilder: BodyBuilder;
  public constructor(apiModel: ApiModel, inputFolder: string) {
    this._apiModel = apiModel;
    this._outputFolder = inputFolder;
    this._emitter = new StandardMarkdownEmitter(apiModel);
    this._pageBuilder = createPageBuilder(CustomDocNodes.configuration, apiModel, (apiItem: ApiItem) => {
      this._writeApiItemPage(apiItem);
    });
  }

  public generate(): Record<string, string> {
    this._writeApiItemPage(this._apiModel);
    return this._pages;
  }

  public generateFiles(): void {
    this.generate(); // todo loop
  }

  private _writeApiItemPage(apiItem: ApiItem): void {
    const page: DocSection = this._pageBuilder((b: ISections): void => {
      b.breadcrumb(this._apiModel);
      b.pageHeading(apiItem);
      b.betaWarning(apiItem);
      b.deprecated(apiItem);
      b.decorators(apiItem);
      b.heritageTypes(apiItem);

      let appendRemarks: boolean = true;

      switch (apiItem.kind) {
        case ApiItemKind.Class:
        case ApiItemKind.Interface:
        case ApiItemKind.Namespace:
        case ApiItemKind.Package:
          b.remarks(apiItem);
          appendRemarks = false;
          break;
      }

      switch (apiItem.kind) {
        case ApiItemKind.Class:
          b.classTable(apiItem as ApiClass);
          break;
        case ApiItemKind.Enum:
          b.enumTable(apiItem as ApiEnum);
          break;
        case ApiItemKind.Interface:
          b.interfaceTable(apiItem as ApiInterface);
          break;
        case ApiItemKind.Constructor:
        case ApiItemKind.ConstructSignature:
        case ApiItemKind.Method:
        case ApiItemKind.MethodSignature:
        case ApiItemKind.Function:
          b.parameterTable(apiItem as ApiParameterListMixin);
          b.throws(apiItem);
          break;
        case ApiItemKind.Namespace:
          b.packageOrNamespace(apiItem as ApiNamespace);
          break;
        case ApiItemKind.Model:
          b.modelTable(apiItem as ApiModel);
          break;
        case ApiItemKind.Package:
          b.packageOrNamespace(apiItem as ApiPackage);
          break;
        case ApiItemKind.Property:
        case ApiItemKind.PropertySignature:
          break;
        case ApiItemKind.TypeAlias:
          break;
        case ApiItemKind.Variable:
          break;
        default:
          throw new Error('Unsupported API item kind: ' + apiItem.kind);
      }

      if (appendRemarks) {
        b.remarks(apiItem);
      }
    });

    const filename: string = path.join(this._outputFolder, getFilenameForApiItem(apiItem));

    const builder: StringBuilder = new StringBuilder();

    builder.append(
      '<!-- Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin -->'
    );

    this._pages[filename] = this._emitter.emit(page, builder, {
      contextApiItem: apiItem,
      onGetFilenameForApiItem: (apiItemForFilename: ApiItem) => {
        return getLinkFilenameForApiItem(apiItemForFilename);
      }
    });
  }
}
