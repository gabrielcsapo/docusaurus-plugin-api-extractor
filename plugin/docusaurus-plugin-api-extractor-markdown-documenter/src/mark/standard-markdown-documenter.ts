import { ApiItem, ApiModel } from '@microsoft/api-extractor-model';
import { StringBuilder, DocSection, TSDocConfiguration, DocNode } from '@microsoft/tsdoc';
import path from 'path';

import { CustomDocNodes } from './nodes';
import { StandardMarkdownEmitter } from './standard-markdown-emitter';
import { getFilenameForApiItem, getLinkFilenameForApiItem } from './builders/file-naming';
import { IDocumenterDelegate, IInternalDocumenterDelegate } from './interfaces';
import { promises as fs } from 'fs';
import { InternalDelegate } from './default-delegate';
import { PrimitiveBuilders } from './builders/primitive-builders';
import { SectionBuilders } from './builders/section-builders';

export class StandardMarkdownDocumenter {
  private _emitter: StandardMarkdownEmitter;
  private _pages: Record<string, string> = {};
  private _configuration: TSDocConfiguration;
  private _delegate: IInternalDocumenterDelegate;

  public constructor(delegate: IDocumenterDelegate);
  public constructor(apiModel: ApiModel, outputPath: string);
  public constructor(...args: unknown[]) {
    let delegate: IInternalDocumenterDelegate;
    if (args.length === 2) {
      if (args[0] instanceof ApiModel && typeof args[1] === 'string') {
        delegate = new InternalDelegate({ apiModel: args[0], outputFolder: args[1] });
      } else {
        throw new Error(`You must pass an APIModel and the outputFolder. Recieved: ${args[0]}\n${args[1]}`);
      }
    } else if (args.length === 1 && args[0] && typeof args[0] === 'object') {
      delegate = new InternalDelegate(args[0] as IDocumenterDelegate);
    } else {
      throw new Error(`You must pass either a custom delegate or the APIModel and outputFolder`);
    }

    this._delegate = delegate;
    this._emitter = new StandardMarkdownEmitter(delegate);
    CustomDocNodes.configuration = delegate.configureTSDoc(CustomDocNodes.configuration);
    this._configuration = CustomDocNodes.configuration;
  }

  public async generate(): Promise<Record<string, string>> {
    this._writeApiItemPage(this._delegate.apiModel);
    const pages: Record<string, string> = this._pages;
    this._pages = {};
    return pages;
  }

  public async generateFiles(): Promise<void> {
    Object.entries(await this.generate()).forEach(async ([filePath, content]) => {
      await fs.writeFile(filePath, content);
    });
  }

  private _writeApiItemPage(apiItem: ApiItem): void {
    const { _configuration: configuration, _delegate: delegate } = this;
    const primitiveBuilders: PrimitiveBuilders = new PrimitiveBuilders(configuration, delegate.apiModel);
    const section: DocSection = primitiveBuilders.section();

    const sectionBuilders: SectionBuilders = new SectionBuilders(
      primitiveBuilders,
      section,
      delegate,
      apiItem,
      this._writeApiItemPage.bind(this)
    );

    delegate.writePage({
      currentPage: section,
      apiItem,
      tsDocConfiguration: configuration,
      append(docNode: DocNode): void {
        section.appendNode(docNode);
      },
      next: (apiItem: ApiItem) => {
        this._writeApiItemPage(apiItem);
      },
      primitives: primitiveBuilders,
      sections: sectionBuilders
    });

    const filename: string = path.join(this._delegate.outputFolder, getFilenameForApiItem(apiItem));

    const builder: StringBuilder = new StringBuilder();

    this._pages[filename] = this._emitter.emit(section, builder, {
      contextApiItem: apiItem,
      onGetFilenameForApiItem: (apiItemForFilename: ApiItem) => {
        return getLinkFilenameForApiItem(apiItemForFilename);
      }
    });
  }
}
