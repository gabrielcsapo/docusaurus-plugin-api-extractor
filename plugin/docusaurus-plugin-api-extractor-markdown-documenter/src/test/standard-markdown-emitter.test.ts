import { ApiModel } from '@microsoft/api-extractor-model';
import { join } from 'path';
import { Documenter } from '../mark/documenter';

it('emits plain strings', () => {
  const model = new ApiModel();
  model.loadPackage(join(__dirname, './fixtures/api-model.custom-types.json'));
  const documenter = new Documenter(model, 'foo');
  expect(documenter.generate()).toMatchSnapshot();
});
