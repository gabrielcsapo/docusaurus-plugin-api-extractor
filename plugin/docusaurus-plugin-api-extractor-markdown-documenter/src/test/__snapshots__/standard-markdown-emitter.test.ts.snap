// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emits plain strings 1`] = `
Object {
  "foo/addon.md": "---
# Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin
id: addon
hide_title: true
title: addon package
---
[Home](./index.md)

## addon package


## Components

|  Components | Description | Import Path |
|  --- | --- | --- |
|  [Stepper](./addon.stepper.md) |  | my-addon/components/stepper |

## Component Signatures

|  Component Signatures | Description | Import Path |
|  --- | --- | --- |
|  [StepperSignature](./addon.steppersignature.md) |  | my-addon/components/stepper |

## Helpers

|  Helpers | Description | Import Path |
|  --- | --- | --- |
|  [sum(a, b)](./addon.sum.md) |  | addon/helpers/sum |

",
  "foo/addon.stepper._constructor_.md": "---
# Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin
id: addon.stepper._constructor_
hide_title: true
title: Stepper.(constructor)
---
[Home](./index.md)

## Stepper.(constructor)

Constructs a new instance of the \`Stepper\` class

**Signature:**

\`\`\`typescript
constructor(args: StepperSignature);
\`\`\`

## Parameters

|  Parameter | Type | Description |
|  --- | --- | --- |
|  args | StepperSignature |  |

",
  "foo/addon.stepper.args.md": "---
# Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin
id: addon.stepper.args
hide_title: true
title: Stepper.args property
---
[Home](./index.md)

## Stepper.args property

**Signature:**

\`\`\`typescript
args: StepperSignature;
\`\`\`
",
  "foo/addon.stepper.md": "---
# Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin
id: addon.stepper
hide_title: true
title: Stepper Component
---
[Home](./index.md)

## Stepper Component

### Import Path: my-addon/components/stepper

**Signature:**

\`\`\`typescript
export default class Stepper<StepperSignature> 
\`\`\`

## Constructors

|  Constructor | Modifiers | Description |
|  --- | --- | --- |
|  [(constructor)(args)](./addon.stepper._constructor_.md) |  | Constructs a new instance of the <code>Stepper</code> class |

## Properties

|  Property | Modifiers | Type | Description |
|  --- | --- | --- | --- |
|  [args](./addon.stepper.args.md) |  | StepperSignature |  |

## Methods

|  Method | Modifiers | Description |
|  --- | --- | --- |
|  [render()](./addon.stepper.render.md) |  |  |

",
  "foo/addon.stepper.render.md": "---
# Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin
id: addon.stepper.render
hide_title: true
title: Stepper.render() method
---
[Home](./index.md)

## Stepper.render() method

**Signature:**

\`\`\`typescript
render(): void;
\`\`\`
**Returns:**

void

",
  "foo/addon.steppersignature.md": "---
# Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin
id: addon.steppersignature
hide_title: true
title: StepperSignature Component Signature
---
[Home](./index.md)

## StepperSignature Component Signature

### Import Path: my-addon/components/stepper

**Signature:**

\`\`\`typescript
export declare type StepperSignature = {
    Element: HTMLDivElement;
    Blocks: {
        default: unknown;
    };
    Arguments: {
        name: string;
        age: number;
    };
};
\`\`\`
",
  "foo/addon.sum.md": "---
# Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin
id: addon.sum
hide_title: true
title: sum() Helper
---
[Home](./index.md)

## sum() Helper

### Import Path: addon/helpers/sum

**Signature:**

\`\`\`typescript
export default function sum(a: number, b: number): number;
\`\`\`

## Parameters

|  Parameter | Type | Description |
|  --- | --- | --- |
|  a | number | left side |
|  b | number | right side |

**Returns:**

number

the sum of the left and right side

## Example

This is a helper that can be used in

\`\`\`ts
import sum from 'my-addon/helpers/sum';

export default <template>{{sum 1 2}}</template>
\`\`\`

",
  "foo/index.md": "---
# Do not edit this file. It is automatically generated by the docusaurus-api-extractor plugin
id: index
hide_title: true
title: API Reference
slug: /
---
[Home](./index.md)

## API Reference

## Packages

|  Package | Description |
|  --- | --- |
|  [addon](./addon.md) |  |

",
}
`;
